File: CS/PS/Site/leetcode/graph/46.2.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function (nums) {
  let visited = Array(nums.length).fill(false);
  let result = [];
  let tmp = [];
  function dfs() {
    if (tmp.length === nums.length) {
      result.push([...tmp]);
      return;
    }
    for (let i = 0; i < nums.length; i++) {
      if (visited[i] === true) continue;
      visited[i] = true;
      tmp.push(nums[i]);
      dfs();
      tmp.pop();
      visited[i] = false;
    }
  }
  dfs()
  return result;
};
console.log(permute([1, 2, 3]));
console.log(permute([0, 1]));

File: CS/PS/Site/leetcode/graph/46.2.py
from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        visited = [False] * len(nums)
        result = []
        tmp = []

        def dfs():
            if len(nums) == len(tmp):
                result.append(tmp[:])
                return
            for v in range(len(nums)):
                if visited[v] == True:
                    continue
                visited[v] = True
                tmp.append(nums[v])
                dfs()
                tmp.pop()
                visited[v] = False
        dfs()
        return result
    print(permute("", [1, 2, 3]))
    print(permute("", [0, 1]))
